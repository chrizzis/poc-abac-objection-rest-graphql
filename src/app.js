import express from 'express'
import path from 'path'
import cors from 'cors'
import attachCurrentUser from './http-middleware/attachCurrentUser.js'
import { apiUrl } from '../config/index.js'
import userRoute from './routes/user.route.js'
import projectRoute from './routes/project.route.js'
import authRoute from './routes/auth.route.js'
const __dirname = path.resolve()

const app = express()

app.use(cors())
app.use(express.urlencoded({ extended: false }))
app.use(express.json())
app.use(express.static(path.join(__dirname, `/public`)))

app.use(authRoute)
app.use(apiUrl, attachCurrentUser)
app.use(apiUrl, userRoute)
app.use(apiUrl, projectRoute)

export default app