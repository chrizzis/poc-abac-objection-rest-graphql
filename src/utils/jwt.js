import jwt from 'jsonwebtoken'
// import dotenv from 'dotenv'
const { PUBLIC_KEY, PRIVATE_KEY } = process.env

const iss = 'jwt-node' // server?
const sub = 'jwt-node' // user.id
const aud = 'jwt-node' // client?

const defaultOptions = {
  issuer: iss,
  subject: sub,
  audience: aud,
  // expiresIn: '24h',
  expiresIn: '30s',
  algorithm: 'RS256',
}

// function expression vs declaration (function generateJwt...) - protect the class name from being reassigned within the class
// but is not hoisted, so any code that calls a function before it is expressed(?) will throw a ReferenceError
// payload = user: { id, username, roles } OR (refreshToken) { exp, sub }
const generateJWT = (payload, quick = false) => {
  let options
  if (payload) {
    if (payload.exp) { // refresh token
      const { expiresIn, subject, ...filteredOptions } = defaultOptions
      options = filteredOptions
    } else { // user token
      const { subject, ...filteredOptions } = defaultOptions
      console.debug(`generateJwt payload should be user.json: ${JSON.stringify(payload)}`)
      console.debug(`payload.id: ${payload.id.toString()}`)
      options = {
        ...filteredOptions,
        subject: payload.id.toString(),
        // subject: payload.id,
      }
      if (quick) {
        options.expiresIn = '10s'
      }
    }
  }
  return jwt.sign(payload, PRIVATE_KEY, options)
}

// TODO: keep track of request IP address origins to help detect token theft
// TODO: currently this is being used to verify user jwts, not refresh tokens
const verifyJWT = (payload) => {
  const { subject, ...filteredOptions } = defaultOptions
  return jwt.verify(payload, PUBLIC_KEY, filteredOptions)
}

export {
  generateJWT,
  verifyJWT,
}