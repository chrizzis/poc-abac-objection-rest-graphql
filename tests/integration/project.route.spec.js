import request from 'supertest'
import app from '../../src/app.js'
import { apiUrl, currentEnv } from '../../config/index.js'
import pool from '../../src/db.js'
import { Model } from 'objection'
import { generateJWT } from '../../src/utils/jwt.js'

// import Knex from 'knex'
// import knexConfigs from '../../knexfile.js'

// const knexConfig = knexConfigs[currentEnv]
// const knex = Knex(knexConfig);

describe("project.route.js", () => {
  let knex
  beforeAll(async () => {
    // knex = Knex(knexConfig);
    knex = pool
    // Model.knex(knex) // not needed for tests to pass, not using objection here...
    // await knex.migrate.rollback();
    // await knex.migrate.latest(); // done in global setup

    // await knex.seed.run(); // TODO: if kept here, can't destroy data in any tests that other tests rely on
  });
  beforeEach(async () => {
    // await knex.seed.run();
  });
  // TODO: this isnt needed b/c globalTeardown
  // afterAll(async () => {
  //   await knex.destroy()
  // })
  describe("GET /projects", () => {
    test("It should return a list of projects filtered by acl for anonymous user", async () => {
      const { body: projects } = await request(app).get(`${apiUrl}/projects`).expect(200);
      // as ac strategies take over, these become obsolete
      // const dbProjects = await knex('Project');
      // expect(projects).toEqual(dbProjects);
      expect(projects).toMatchSnapshot();
    });
    test("It should return a list of projects filtered by acl for registered user", async () => {
      const seedUsers = await knex('User');
      const seedUser = seedUsers[0]
      const token = generateJWT(seedUser)
      const { body: projects } = await request(app)
        .get(`${apiUrl}/projects`)
        .set('Authorization', `Bearer ${token}`)
        .expect(200);
      expect(projects).toMatchSnapshot();
    });
  });
  describe("GET /projects/:id", () => {
    test("It should return a project for an authorized (owner) user", async () => {
      const seedUsers = await knex('User');
      const seedUser = seedUsers[0]
      const token = generateJWT(seedUser)
      const dbProjects = await knex('Project');
      const { id } = dbProjects[0]
      const { body: project } = await request(app)
        .get(`${apiUrl}/projects/${id}`)
        .set('Authorization', `Bearer ${token}`)
        .expect(200);
      expect(project).toMatchSnapshot();
    });
    test("It should return 401 for an unauthenicated user", async () => {
      const dbProjects = await knex('Project');
      const { id } = dbProjects[0]
      await request(app).get(`${apiUrl}/projects/${id}`).expect(401);
    });
    test("It should return 403 for an unauthorized user", async () => {
      const seedUsers = await knex('User');
      const seedUser = seedUsers[1]
      const token = generateJWT(seedUser)
      const dbProjects = await knex('Project');
      const { id } = dbProjects[0]
      await request(app)
        .get(`${apiUrl}/projects/${id}`)
        .set('Authorization', `Bearer ${token}`)
        .expect(403);
    });
  });
});