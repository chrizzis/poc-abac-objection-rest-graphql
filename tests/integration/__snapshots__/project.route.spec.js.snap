// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`project.route.js GET /projects It should return a list of projects filtered by acl for anonymous user 1`] = `
Array [
  Object {
    "id": 4,
    "ownerId": 1,
    "published": true,
    "title": "Published project by Alice",
  },
  Object {
    "id": 5,
    "ownerId": 2,
    "published": true,
    "title": "Published project by Bob",
  },
  Object {
    "id": 6,
    "ownerId": 3,
    "published": true,
    "title": "Published project by Eve",
  },
]
`;

exports[`project.route.js GET /projects It should return a list of projects filtered by acl for registered user 1`] = `
Array [
  Object {
    "id": 1,
    "ownerId": 1,
    "ownerOnly": "only alice can see this",
    "published": false,
    "title": "Project by Alice",
  },
  Object {
    "id": 4,
    "ownerId": 1,
    "ownerOnly": "only alice can see this",
    "published": true,
    "title": "Published project by Alice",
  },
  Object {
    "id": 5,
    "ownerId": 2,
    "published": true,
    "title": "Published project by Bob",
  },
  Object {
    "id": 6,
    "ownerId": 3,
    "published": true,
    "title": "Published project by Eve",
  },
]
`;

exports[`project.route.js GET /projects/:id It should return a project for an authorized (owner) user 1`] = `
Object {
  "id": 1,
  "ownerId": 1,
  "ownerOnly": "only alice can see this",
  "published": false,
  "title": "Project by Alice",
}
`;
